reconcile/terraform_cloudflare_dns.py

```diff
@@ -354,34 +287,69 @@ def build_clients(
+
+        if cf_acct_name in cf_accounts:
+            cf_account = cf_accounts[cf_acct_name]
         else:
-            cf_acct_config = get_cf_acct_config(cf_acct, secret_reader)
-            cf_acct_configs[cf_acct_name] = cf_acct_config
-        aws_acct = cf_acct.terraform_state_account
-        aws_backend_config = create_backend_config(
-            secret_reader,
-            aws_acct,
-            cf_acct,
-            zone.identifier,
-            integration_name=integration_name,
+            cf_account = CloudflareAccount(
+                cf_acct_name,
+                zone.account.api_credentials,
+                zone.account.enforce_twofactor,
+                zone.account.q_type,
+                zone.account.provider_version,
+            )
+            cf_accounts[cf_acct_name] = cf_account
+
+        tf_state = zone.account.terraform_state_account.terraform_state
+        if not tf_state:
+            raise ValueError(
+                f"AWS account {zone.account.terraform_state_account.name} cannot be used for Cloudflare "
+                f"account {cf_account.name} because it does not define a Terraform state "
+            )
+        bucket = tf_state.bucket
+        region = tf_state.region
+        integrations = tf_state.integrations
+
+        if not bucket:
+            raise InvalidTerraformState("Terraform state must have bucket defined")
+        if not region:
+            raise InvalidTerraformState("Terraform state must have region defined")
+
+        integration = None
+        for i in integrations:
+            if i.integration.replace("-", "_") == qontract_integration:
+                integration = i
+                break
+
+        if not integration:
+            raise IntegrationUndefined(
+                "Must declare integration name under Terraform state in app-interface"
+            )
+
+        tf_state_s3 = TerraformStateS3(
+            zone.account.terraform_state_account.automation_token,
+            bucket,
+            region,
+            Integration(integration.integration.replace("-", "_"), integration.key),
         )
 
         rps = get_cloudflare_provider_rps(zone.records)
 
-        ts_config = create_cloudflare_terrascript(
-            account_config=cf_acct_config,
-            backend_config=aws_backend_config,
-            provider_version=cf_acct.provider_version,
-            provider_rps=rps,
-            is_managed_account=False,
+        client = TerrascriptCloudflareClientFactory.get_client(
+            tf_state_s3,
+            cf_account,
+            DNSZoneShardingStrategy(cf_account, zone.identifier),
+            secret_reader,
+            False,
+            rps,
         )
 
-        ts_client = TerrascriptCloudflareClient(ts_config)
-        clients.append((f"{cf_acct.name}-{zone.identifier}", ts_client))
+        try:
+            cf_clients.register_client(f"{cf_account.name}-{zone.identifier}", client)
```

Comment:
Great question. This is because the zone based sharding. To be able to differentiate state for each zone. Let me know if that makes sense :)

Reply:
With the new cloudflare client implementation, state based sharding is done using `DNSZoneShardingStrategy` through `get_object_key` method which will be used to create S3 object.

Reply:
btw don't block on this comment, since we are keeping the behavior same with the refactor and also I'm lot less familiar with the design decisions made with dns integration.

Reply:
Right, but when adding the spec, each externalResource still need to be differentiated. If a Cloudflare account contain multiple zones, and all the client for each zone are registered under the same account, when calling add_spec on each client, how does it know how to add the resource to different zone?

Reply:
ah! that makes sense. thanks for the explanation.